Add EVM integration to docs (#202)
timewave-computer/valence-protocol: docs/src/authorizations_processors/execution_environment_differences.md

# Execution Environment Differences

Depending on the type of `Execution Environment` being used, the behavior of the `Processor` may vary. In this section we will describe the main differences in how the `Processor` behaves in the different `Execution Environments` that we support.

### Execution Success

> During the execution of a `Message Batch`, the `Processor` will execute each function of the subroutine of that batch. If the execution for a specific function fails, we will consider the execution failed in case of `Atomic` batches, and we will stop the execution of the next function in case of `NonAtomic` batches.

Reviewer: ```suggestion
During the execution of a `Message Batch`, the `Processor` will execute each function of the subroutine of that batch. If the execution for a specific function fails, batch execution is considered a failure. For `Atomic` batches, execution is terminated upon any message execution failure. Execution continues to the next function in case of message execution failure within `NonAtomic` batches.
```

Author: Not sure of this: Execution continues to the next function in case of message execution failure within `NonAtomic` batches. If a function fails in NonAtomic batches we don't continue to the next one, we stop executing. Maybe I misunderstood this change

> Currently, in the `CosmWasm` Execution Environment, a function fails if the `CosmWasm` contract that we are targeting doesn't exist, if the `entry point` of that contract doesn't exist, or if the execution of the contract fails for any reason. On the contract, in the `EVM` Execution Environment, a function only fails if the contract explicitly fails or reverts.

Reviewer: ```suggestion
Currently, in the `CosmWasm` execution environment, a function fails if the `CosmWasm` contract that we are targeting doesn't exist, if the `entry point` of that contract doesn't exist, or if the execution of the contract fails for any reason. On the contract, in the `EVM` execution environment, a function only fails if the contract explicitly fails or reverts.
```

> To mitigate the differences in behavior between these two `Execution Environments`, we have added a check in the `EVM` Processor version to consider the execution failed if the contract doesn't exist by explicitly checking if the contract indeed exists. We've also added a revert in our `EVM` libraries if the execution of the contract enters the fallback function, which is not allowed in our system. Nevertheless, since the `Processors` are not restricted to `Valence Libraries` but can call any contract, we can't guarantee that the contract targeted will fail if an entry point doesn't exist, because the fallback function might not be defined or might not revert.

Reviewer: ```suggestion
To mitigate the differences in behavior between these two execution environments, a an `EVM` Processor version check was included to check if the contract indeed exists and fail execution if the contract does not exist. Behavior was also added in the `EVM` libraries to revery if the execution of the contract enters the fallback function, which is not allowed in the system. Nevertheless, since `Processors` are not restricted to `Valence Libraries` but can call any contract, no guarantee can be made that the contract targeted will fail if an entry point does not exist, because the fallback function might not be defined or might not revert.
```

> Therefore, whereas in `CosmWasm` executions, a contract will always fail if the entry point doesn't exist, in `EVM` executions, this is not necessarily the case. This is a difference that the owner of the program needs to be taken into account when designing and creating this program.

Reviewer: ```suggestion
In `CosmWasm` execution of a contract will always fail if the entry point does not exist. However, for `EVM` execution, this is not necessarily the case. This is a difference that the owner of the program mist take into account when designing and creating their program.
```

> ***In summary***: if a function of the subroutine targets a contract that is NOT a `Valence Library` AND the entry point of that contract doesn't exist AND the fallback function is either not defined or doesn't revert, the execution of that function will be considered successful in the `EVM` Execution Environment while it wouldn't be in the `CosmWasm` Execution Environment equivalent.

Reviewer: ```suggestion
***In summary***: if a function of the subroutine targets a contract that is NOT a `Valence Library` AND the entry point of that contract does not exist AND the fallback function is either not defined or doesn't revert, the execution of that function will be considered successful in the `EVM` execution environment while it wouldn't be in the `CosmWasm` Execution Environment equivalent.
```
This is a *very* complicated sentence. Can we bullet point this or something?

Author: Bullet pointed it
